pkg/kubelet/cm/dra/plugin/plugin.go

```diff
@@ -18,94 +18,140 @@ package plugin
+		ctx:        klog.NewContext(context.TODO(), klog.LoggerWithName(klog.TODO(), "DRA registration handler")),
+		kubeClient: kubeClient,
+		restConfig: restConfig,
+		getNode:    getNode,
+	}
 
-	// If kubelet ever gets an API for stopping registration handlers, then
-	// that would need to be hooked up with stopping the controller.
-	handler.controller = startNodeResourcesController(context.TODO(), kubeClient, getNode)
+	// When kubelet starts up, no DRA driver has registered yet. None of
+	// the drivers are usable until they come back, which might not happen
+	// at all. Therefore it is better to not advertise any local resources
+	// because pods could get stuck on the node waiting for the driver
+	// to start up.
+	//
+	// This has to run in the background.
+	go handler.wipeResourceSlices("")
 
 	return handler
 }
 
+// wipeResourceSlices deletes ResourceSlices of the node, optionally just for a specific driver.
+func (h *RegistrationHandler) wipeResourceSlices(pluginName string) {
+	if h.restConfig == nil {
+		return
+	}
+	ctx := h.ctx
+	logger := klog.FromContext(ctx)
+	backoff := wait.Backoff{
+		Duration: time.Second,
+		Factor:   2,
+		Jitter:   0.2,
+		Cap:      5 * time.Minute,
+		Steps:    100,
+	}
+	discoveryCache := memory.NewMemCacheClient(h.kubeClient.Discovery())
+	restMapper := restmapper.NewDeferredDiscoveryRESTMapper(discoveryCache)
+	gk := resourceapi.SchemeGroupVersion.WithKind("ResourceSlice").GroupKind()
+	dynClient, err := dynamic.NewForConfig(h.restConfig)
+	if err != nil {
+		logger.Error(err, "Creating dynamic client for REST config failed, not deleting ResourceSlices")
+		return
+	}
+
+	// Error logging is done inside the loop. Context cancellation doesn't get logged.
+	_ = wait.ExponentialBackoffWithContext(ctx, backoff, func(ctx context.Context) (bool, error) {
+		node, err := h.getNode()
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		if err != nil {
+			logger.Error(err, "Unexpected error checking for node")
+		}
+		fieldSelector := fields.Set{"nodeName": node.Name}
+		if pluginName != "" {
+			fieldSelector["driverName"] = pluginName
+		}
+
+		mapping, err := restMapper.RESTMapping(gk)
+		if err != nil {
+			// https://github.com/kubernetes/kubernetes/blob/b722d017a34b300a2284b890448e5a605f21d01e/staging/src/k8s.io/client-go/restmapper/discovery.go#L291
+			// does not invalidate the cache on cache misses because the cache is "fresh".
+			// We need to detect when the API comes back, so we have to invalidate ourselves.
+			restMapper.Reset()
+
+			// Not found or some other error. This could get resolved by an API server
+			// restart, so we keep trying.
+			logger.V(5).Info("Looking up ResourceSlice REST mapping failed, retrying", "err", err)
+			return false, nil
+		}
+
+		resourceClient := dynClient.Resource(mapping.Resource)
+		err = resourceClient.DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{FieldSelector: fieldSelector.String()})
```

Comment:
if the only call we're making is fieldSelector-scoped DeleteCollection call, we don't need a dynamic client for that, right?

Reply:
Correct.